src/witness/witnessUtil.ml

```diff
@@ -79,7 +79,61 @@ struct
       emit_after_lock
     else
       emit_other
+
+  let find_syntactic_loop_head n =
+    let prevs = Cfg.prev n in
+    List.find_map (fun (edges, prev) ->
+        let stmts = Cfg.skippedByEdge prev edges n in
+        List.find_map (fun s ->
+            match s.GoblintCil.skind with
+            | Loop (_, loc, _, _, _) -> Some loc
+            | _ -> None
+          ) stmts
+      ) prevs
+end
+
+module YamlInvariant (FileCfg: MyCFG.FileCfg) =
+struct
+  include Invariant (FileCfg)
+
+  let is_stub_node n =
+    let fundec = Node.find_fundec n in
+    Cil.hasAttribute "goblint_stub" fundec.svar.vattr
+
+  let location_location (n : Node.t) = (* great naming... *)
+    match n with
+    | Statement s ->
+      let {loc; _}: CilLocation.locs = CilLocation.get_stmtLoc s in
+      if not loc.synthetic && is_invariant_node n && not (is_stub_node n) then (* TODO: remove is_invariant_node? *)
```

Comment:
What's up with the TODO?

Reply:
This is another side of https://github.com/goblint/analyzer/pull/1372#discussion_r1528301516. I'll expand the comment on it.